polkadot/node/network/statement-distribution/src/v2/grid.rs

```diff
@@ -908,6 +913,14 @@ struct MutualKnowledge {
 	/// `Some` only if we have advertised, acknowledged, or requested the candidate
 	/// from them.
 	local_knowledge: Option<StatementFilter>,
+	/// Knowledge peer sent to us, this is different from
+	/// `local_knowledge` and `remote_knowledge`, through the fact that includes only
+	/// statements that we received from peer while the other two include knowledge from
+	/// what we have sent to the peer and he sent back.
+	/// This is needed because sometimes we have races where peers send us statements that
```

Comment:
I might be missing something, but .... Is there any reason we cannot handle the race by accepting statements for candidates we have already sent, but just once. That can be as easy setting only `local_knowledge` to `Some` when have advertised, acknowledged, or requested the candidate from the peer, but not the `remote_knowledge` that we'd update once the they advertise it once.

Reply:
Let me describe the flow a bit better:
1. When we send the manifest to a peer we set `local_knowledge` and that represents all the statements that we already have and that we are going to send to peer.
2. When receive the manifest we set `remote_knowledge` and that represents all the statements that peer already has.


Now, after this step we can circulate new statements we received from peers and we do that and we call `sent_or_received_direct_statement` which set both the local knowledge and the remote knowledge see here: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/node/network/statement-distribution/src/v2/grid.rs#L1020, hence why local_knowledge will be set to true and we have no way to know if the peer sent us the message 0, 1 or 2 times.

Now a solution would be probably to try and break sent_or_received_direct_statement in 2 parts `sent` and `received` however that would be a more intrusive solution and would be easy to get things wrong while doing, which I didn't think was worth the effort for this fix which is minor anyway.

But, I can try to do that if you think this fix is a bit too ugly.